File: src/main/java/com/azbj/proposalde3/service/ControlService.java
package com.azbj.proposalde3.service;

import com.azbj.proposalde3.repository.ControlRepository;
import com.azbj.proposalde3.model.Control;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Map;

@Service
public class ControlService {

    @Autowired
    private ControlRepository controlRepository;

    public byte[] getImage(String policyType) {
        return controlRepository.findImageByPolicyType(policyType);
    }

    public Control updatePolicyHolderDetails(Control policyHolderDetails) {
        return controlRepository.save(policyHolderDetails);
    }

    public void clearDetails() {
        controlRepository.deleteAll();
    }

    public Control approvePanDetails(Control panApprovalDetails) {
        return controlRepository.save(panApprovalDetails);
    }

    public boolean validateEmployeeId(String employeeId, String fscCode) {
        return controlRepository.validateEmployeeId(employeeId, fscCode);
    }

    public byte[] loadImage() {
        return controlRepository.loadImageData();
    }

    public void updateAssignmentFlag(boolean flag) {
        controlRepository.updateAssignmentFlag(flag);
    }

    public Map<String, Object> verifyAadhaarDetails(Map<String, Object> customerProfileDetails) {
        return controlRepository.findAadhaarDetails(customerProfileDetails);
    }

    public Control saveAMLInformation(Control amlInfo) {
        return controlRepository.save(amlInfo);
    }

    public String getProofDescription(String proofType) {
        return controlRepository.findProofDescriptionByType(proofType);
    }

    public void processEnrichmentDetails(String applicationNo, String callFormName, String proposalNo, String laName) {
        // Check if a parameter list named 'Param1' exists.
        // If it exists, destroy and recreate it.
        // Add the following parameters to the newly created parameter list:
        // `APPLICATION_NO` with the value of the application number.
        // `CALL_FORM_NAME` with the value 'DEQC'.
        // `PROPOSAL_NO` with a null value.
        // `LA_NAME` with the concatenated insured person's name.
        // Call a form named 'AZBJ_DATA_ENRICH_FORM' with the created parameter list.
        // Handle any errors that occur during this process and return an appropriate error message.
    }

    public boolean validatePan(String documentId) {
        return controlRepository.validatePan(documentId);
    }

    public void createParameterList(Map<String, Object> params) {
        // Check if a parameter list named 'Param1' already exists. If it does, destroy it.
        // Create a new parameter list named 'Param1'.
        // Add the provided parameters to the parameter list.
    }

    public void callForm(String formName, Map<String, Object> params) {
        // Call the specified form with the provided parameters.
    }

    public String getScrutinyNumber(String applicationNumber) {
        return controlRepository.findScrutinyNumberByApplicationNumber(applicationNumber);
    }

    public boolean checkEiaDetails(String applicationNumber) {
        return controlRepository.findEiaDetailsByApplicationNumber(applicationNumber);
    }

    public void navigateToAMLBlock() {
        // Navigate to the 'AML' block and iterate through all records to find the value of 'PAN Card' if the 'proof_type' is 'PC'.
    }

    public void incrementCounter() {
        // Increment the counter `pk_vars.v_document_de_cnt`.
    }

    public void callFinancialDocumentForm(Map<String, Object> paramList) {
        // Call the form 'AZBJ_FINANCIAL_DOCUMENT' with the given parameter list.
    }

    public Map<String, Object> retrievePassportDetails() {
        return controlRepository.findPassportDetails();
    }

    public void openSameBankDetails(String bankAccountNumber, String ifscCode) {
        // Validate the bank account number and IFSC code.
        // If valid, proceed to open the 'AZBJ_SAME_BANK_DETAILS' form.
        // Handle any exceptions that occur during this process.
    }

    public Map<String, Object> getDrivingLicenseDetails() {
        return controlRepository.findDrivingLicenseDetails();
    }

    public boolean verifyVoterId() {
        return controlRepository.fetchAgeProof();
    }

    public boolean verifyPanStatus(String status) {
        // Verify the PAN status and update the database accordingly.
        return true;
    }

    public void populateQuestions(Map<String, Object> partnerDetails, String applicationNumber) {
        // Fetch questions based on partner details and application number, delete existing records if member number matches partner number, create new records with populated questions, and delete specific questions based on predefined conditions.
    }

    public boolean validatePanCard(String panCardNumber) {
        return controlRepository.validatePanCard(panCardNumber);
    }

    public boolean saveDuplicateContactDetails(Control controlModel) {
        return controlRepository.saveDuplicateContactDetails(controlModel);
    }

    public boolean validateAmount(double amount) {
        return amount > 0;
    }

    public int calculateNomineeAge(java.util.Date nomineeDOB) {
        // Calculate the nominee's age based on the date of birth and the current date.
        return 0;
    }

    public boolean verifyAndInsertEiaDetails(String applicationNo, String accountType) {
        // Check if an EIA exists for the given application number and account type.
        // If it exists, verify the existence of the applicant's details in the DEQC table, retrieve internal values for state, nation, and bank, and insert the applicant's details into the DEQC table if they do not already exist.
        return true;
    }

    public boolean saveSelectedOption(String selectedOption) {
        return controlRepository.saveSelectedOption(selectedOption);
    }

    public Map<String, Object> getNomineeDetails(String applicationNo) {
        return controlRepository.findNomineeDetails(applicationNo);
    }

    public boolean validateProductCode(String productCode) {
        return controlRepository.validateProductCode(productCode);
    }

    public String getContractId(String applicationNumber) {
        return controlRepository.getContractId(applicationNumber);
    }

    public void deleteExistingSignatureVerification(String contractId) {
        controlRepository.deleteExistingSignatureVerification(contractId);
    }

    public void insertNewSignatureVerification(String contractId, String signatureDecision) {
        controlRepository.insertNewSignatureVerification(contractId, signatureDecision);
    }

    public boolean validatePAN(String panNumber) {
        return controlRepository.validatePAN(panNumber);
    }

    public Map<String, Object> fetchCRPFDefenceQuestions(Map<String, Object> userDetails) {
        return controlRepository.findCRPFDefenceQuestions(userDetails);
    }

    public boolean verifyAadharCard(String aadharNumber) {
        return controlRepository.checkAadharCard(aadharNumber);
    }

    public void callFormWithParams(Map<String, Object> params) {
        // Process the parameter list.
        // Call the form 'AZBJ_SOLN_FORM_CALL' with the parameter list.
    }

    public boolean verifyAadhar(String aadharNumber) {
        return controlRepository.findByAadharNumber(aadharNumber);
    }

    public String generateBIReport() {
        // Retrieve the BI link from the database, concatenate it with the control.bi field, and update the counters and URL.
        return "";
    }

    public boolean saveApplicationDetails(String applicationNo, String moduleFlag, String topIndicator, Map<String, Object> sipControlDetails) {
        // Check for existing records, update them if found, and insert new records based on the provided details.
        // Handle transaction management and commit the transaction after all operations are completed.
        return true;
    }

    public boolean verifyPanAadharLink(String applicationNumber, String transactionType) {
        // Check if the desktop flag is 'N'.
        // Fetch the contract ID from the azbj_batch_items table using the application number and transaction type 'FRP'.
        // Determine the value of v_ip_no based on the CHK_INSURED flag and the age of the user.
        // Navigate to the azbj_qc_questions block and iterate through the records to check if the question ID is 80 and if the answer is 'Y'. If so, set the mandatory flag to 'Y'.
        // Determine the nationality and IP nationality flags based on the user's nationality.
        // Check if the age proof is 'PP' (passport proof) and set the passport proof flag accordingly. If the passport proof flag is 'N', navigate to the AML block and iterate through the records to check if any proof type is 'PP'.
        // If the mandatory flag is 'Y', create a parameter list with various parameters and call the AZBJ_INCOMETAX_QUEST_DTLS form with these parameters.
        // If the mandatory flag is not 'Y', display a warning message.
        return true;
    }

    public boolean validatePAN(String panNumber) {
        return controlRepository.validatePAN(panNumber);
    }

    public Map<String, Object> getEmployerNameValues() {
        return controlRepository.fetchEmployerNameValues();
    }

    public boolean validateAadhaar(Map<String, Object> aadhaarDetails) {
        return controlRepository.findByAadhaar(aadhaarDetails);
    }
}

File: src/app/services/control.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ControlModel } from '../models/control.model';

@Injectable({
  providedIn: 'root'
})
export class ControlService {
  constructor(private http: HttpClient) {}

  getImage(policyType: string): Observable<any> {
    return this.http.get(`/api/image?policyType=${policyType}`);
  }

  getReceiptDetails(): Observable<any> {
    return this.http.get('/api/receipts');
  }

  updatePolicyHolderDetails(policyHolderDetails: any): any {
    // Logic to update policy holder details
    return policyHolderDetails;
  }

  clearDetails(): any {
    // Logic to clear policy holder details
    return {};
  }

  approvePanDetails(panApprovalDetails: ControlModel): Observable<any> {
    return this.http.post('/api/approvePanDetails', panApprovalDetails);
  }

  validateEmployeeId(employeeId: string, fscCode: string): Observable<any> {
    return this.http.post('/api/validateEmployeeId', { employeeId, fscCode });
  }

  loadImage(): Observable<any> {
    return this.http.get('/api/loadImage');
  }

  updateMailingAddress(residenceAddress: any): any {
    // Logic to update mailing address
    return residenceAddress;
  }

  updateAssignmentFlag(flag: boolean): Observable<any> {
    return this.http.post('/api/control/updateAssignmentFlag', { flag });
  }

  checkAndCreateParamList(paramName: string, agentCode: string): any {
    // Logic to check and create parameter list
    return { paramName, agentCode };
  }

  openForm(formName: string, paramList: any): any {
    // Logic to open form
    return { formName, paramList };
  }

  getAadhaarDetails(customerProfileDetails: any): Observable<any> {
    return this.http.post('/api/retrieveAadhaarDetails', customerProfileDetails);
  }

  checkAndCreateParamListForRCU(): any {
    // Logic to check and create parameter list for RCU
    return {};
  }

  callFormForRCU(paramList: any): any {
    // Logic to call form for RCU
    return paramList;
  }

  saveAMLInformation(amlInfo: any): Observable<any> {
    return this.http.post('/api/aml', amlInfo);
  }

  getProofDescription(proofType: string): Observable<any> {
    return this.http.get('/api/aml/proof-description', { params: { proofType } });
  }

  getControlFlag(): Observable<any> {
    return this.http.get('/api/control-flag');
  }

  getEnrichmentDetails(applicationNo: string, callFormName: string, proposalNo: string | null, laName: string): Observable<any> {
    return this.http.post('/api/enrichment-details', { applicationNo, callFormName, proposalNo, laName });
  }

  validatePan(documentId: string): Observable<any> {
    return this.http.post('/api/validatePan', { documentId });
  }

  createParameterList(params: { [key: string]: any }): any {
    // Logic to create parameter list
    return params;
  }

  callForm(formName: string, params: { [key: string]: any }): any {
    // Logic to call form
    return { formName, params };
  }

  getScrutinyNumber(applicationNumber: string): Observable<any> {
    return this.http.get(`/api/control/scrutiny-number/${applicationNumber}`);
  }

  checkEiaDetails(applicationNumber: string): Observable<any> {
    return this.http.get(`/api/control/eia-details/${applicationNumber}`);
  }

  getLOVData(documentType: string, partnerType: string): Observable<any> {
    return this.http.get('/api/aml/lov-data', { params: { documentType, partnerType } });
  }

  getSearchResults(): Observable<any> {
    return this.http.get('/api/search-results');
  }

  fetchDuplicateContacts(): Observable<any> {
    return this.http.get('/api/duplicate-contacts');
  }

  createParameterListForDocumentDE(listName: string): any {
    // Logic to create parameter list for Document DE
    return listName;
  }

  addParameter(listName: string, paramName: string, paramValue: any): any {
    // Logic to add parameter
    return { listName, paramName, paramValue };
  }

  navigateToAMLBlock(): any {
    // Logic to navigate to AML block
    return {};
  }

  incrementCounter(): any {
    // Logic to increment counter
    return {};
  }

  callFinancialDocumentForm(paramList: any): any {
    // Logic to call financial document form
    return paramList;
  }

  getPassportDetails(): Observable<any> {
    return this.http.get('/api/passport-details');
  }

  openSameBankDetails(bankAccountNumber: string, ifscCode: string): any {
    // Logic to open same bank details
    return { bankAccountNumber, ifscCode };
  }

  getPartners(): any {
    // Logic to get partners
    return ['5', '8', '10', 'SL'];
  }

  storeSelectedPartner(selectedPartner: string): any {
    // Logic to store selected partner
    return selectedPartner;
  }

  getDrivingLicenseDetails(): Observable<any> {
    return this.http.get('/api/driving-license-details');
  }

  fetchDataBasedOnCheckbox(checkboxState: boolean): any {
    // Logic to fetch data based on checkbox
    return checkboxState;
  }

  updatePartnerType(partnerType: string): any {
    // Logic to update partner type
    return partnerType;
  }

  verifyVoterId(): Observable<any> {
    return this.http.get('/api/verifyVoterId');
  }

  verifyPanStatus(status: string): Observable<any> {
    return this.http.post('/api/verifyPanStatus', { status });
  }

  populateQuestions(partnerDetails: any, applicationNumber: string): Observable<any> {
    return this.http.post('/api/populateQuestions', { partnerDetails, applicationNumber });
  }

  validateBIDateService(biDate: string, proposalSignDate: string): boolean {
    // Logic to validate BI Date
    return biDate <= proposalSignDate;
  }

  populateCarequote(biNumber: string): any {
    // Logic to populate carequote
    return biNumber;
  }

  validatePanCard(panCardNumber: string): Observable<any> {
    return this.http.post('/api/validatePanCard', { panCardNumber });
  }

  saveDuplicateContactDetails(controlModel: ControlModel): Observable<any> {
    return this.http.post('/api/saveDuplicateContactDetails', controlModel);
  }

  validateAddress(addressDetails: any): boolean {
    // Logic to validate address
    return addressDetails.includes('LEH') || addressDetails.includes('JAMMU AND KASHMIR');
  }

  validateBIReceivedDate(biReceivedDate: string, biDate: string, instab: string, proptype: string, fscCode: string, productCode: string): string {
    if (biReceivedDate < biDate) {
      if (instab === 'INSTAB' && proptype === 'I' && fscCode !== '2000003060' && productCode !== 321) {
        return 'BI Received date should not be less than BI Date.';
      } else {
        return 'BI Received date should not be less than BI Date.';
      }
    }
    return '';
  }

  saveData(): any {
    // Logic to save data
    return {};
  }

  updateUIElements(selectedOption: string): any {
    if (selectedOption === 'CORRECT_NUMBER') {
      return { approveButton: true, supervisorCheck: true, saveButton: false };
    } else {
      return { approveButton: false, supervisorCheck: false, saveButton: true };
    }
  }

  validateAmount(amount: number): Observable<any> {
    return this.http.post('/api/validateAmount', { amount });
  }

  calculateNomineeAge(nomineeDOB: Date): Observable<any> {
    return this.http.post('/api/calculateNomineeAge', { nomineeDOB });
  }

  saveSelectedOption(selectedOption: string): Observable<any> {
    return this.http.post('/api/saveSelectedOption', { selectedOption });
  }

  getNomineeDetails(applicationNo: string): Observable<any> {
    return this.http.get(`/api/nominee-details/${applicationNo}`);
  }

  validateProductCode(productCode: string): Observable<any> {
    return this.http.get(`/api/validateProductCode/${productCode}`);
  }

  verifyEKYC(aadhaarNumber: string, moduleName: string): any {
    // Logic to verify E-KYC
    return { aadhaarNumber, moduleName };
  }

  checkSignatureDecision(applicationNumber: string, signatureDecision: string): Observable<any> {
    return this.http.post('/api/verifySignatureDecision', { applicationNumber, signatureDecision });
  }

  setSelectedOption(selectedOption: string): any {
    // Logic to set selected option
    return selectedOption;
  }

  getSelectedOption(): any {
    // Logic to get selected option
    return {};
  }

  validatePAN(panNumber: string): Observable<any> {
    return this.http.post('/api/validatePAN', { panNumber });
  }

  authenticateSupervisor(supervisorCredentials: any): boolean {
    // Logic to authenticate supervisor
    return true;
  }

  validatePolicySelection(): boolean {
    // Logic to validate policy selection
    return true;
  }

  fetchCRPFDefenceQuestions(userDetails: any): Observable<any> {
    return this.http.post('/api/getCRPFDefenceQuestions', userDetails);
  }

  fetchQuestions(userDetails: any): Observable<any> {
    return this.http.post('/api/fetchQuestions', userDetails);
  }

  verifyAadharCard(aadharNumber: string): Observable<any> {
    return this.http.post('/api/verifyAadharCard', { aadharNumber });
  }

  callFormWithParams(params: any): Observable<any> {
    return this.http.post('/api/callFormWithParams', params);
  }

  verifyAadhar(aadharNumber: string): Observable<any> {
    return this.http.post('/api/verifyAadhar', { aadharNumber });
  }

  saveRelationship(relationship: string): Observable<any> {
    return this.http.post('/api/saveRelationship', { relationship });
  }

  generateBIReport(): Observable<any> {
    return this.http.get('/api/generateBIReport');
  }

  getEmployerNameValues(): Observable<any> {
    return this.http.get('/api/fetchEmployerNameValues');
  }

  validateAadhaar(aadhaarDetails: any): Observable<any> {
    return this.http.post('/api/validateAadhaar', aadhaarDetails);
  }
}
